### PR TITLE
Prompt for ICE server fallback permission

### PR DESCRIPTION
**Pull request by jryans**
When the homeserver does not have any ICE servers configured, this will cause a prompt to appear context of either placing a call or when a call fails. The user's choice is saved on the local device and can be changed in Settings. The fallback ICE server is only used if the user allows it. The dialog also recommends notifying the homeserver admin to rectify the issue.

2019-08-15 at 15 03

Fixes https://github.com/vector-im/riot-web/issues/10173
Depends on https://github.com/matrix-org/matrix-js-sdk/pull/1015

### CODE REVIEW

**Comment by jryans** [comment]
Added @nadonomy for design review, as I made up this text myself, so perhaps he has some better wording.

**Comment by turt2live** [comment]
what doesn't lgtm is that the tests are broken for `Homeserver not configured to support calls` showing up.

**Comment by jryans** [comment]
> what doesn't lgtm is that the tests are broken for `Homeserver not configured to support calls` showing up.

Ah yeah... 😓 I'll look around for the right place to make the tests work around this.

File: src/components/structures/MatrixChat.js [hunk 1]
```diff
@@ -1370,6 +1370,37 @@ export default React.createClass({
...
+                        "Your IP address would be shared with this fallback server " +
+                        "only if you agree and later place or receive a call.",
+                        null, { code },
+                    )}</p>
+                </div>,
+                button: _t('Allow Fallback'),
+                cancelButton: _t('Dismiss'),
+                onFinished: (confirmed) => {
+                    if (confirmed) {
+                        cli.setFallbackICEServerAllowed(true);
```
Comment (jryans) [inline src/components/structures/MatrixChat.js]: This is intended to also be saved as a client only setting.

**Comment by jryans** [comment]
> Do we need terms of service/privacy policy on the fallback TURN server?

I wasn't sure either. After some review, since the only data exposed to the fallback STUN server, and that is called out explicitly in the permission dialog, it has been agreed that we don't need additional documents for this case.

File: src/settings/Settings.js [hunk 2]
```diff
@@ -372,4 +372,10 @@ export const SETTINGS = {
         default: false,
         controller: new LowBandwidthController(),
     },
+    "fallbackICEServerAllowed": {
+        supportedLevels: LEVELS_DEVICE_ONLY_SETTINGS,
+        displayName: _td("Allow fallback call assist server turn.matrix.org"),
```
Comment (turt2live) [inline src/settings/Settings.js]: This might need ab it of wordsmithing - it's moderately unclear what this does in isolation.

**Comment by turt2live** [comment]
fwiw this could also be implemented as a dedicated SettingController to deal with the MatrixClient stuff (both reading and writing). It might not be worth the complexity though.

**Comment by jryans** [comment]
After discussion with @nadonomy, we'll use the following dialog text for the prompt:

>You cannot make calls
>
>Your homeserver (`test.convolv.es`) isn’t configured to make voice or video calls. Ask your homeserver admin to set up a TURN server to enable them.
>
>Alternatively, you can make voice & video calls using our public STUN server (hosted at `turn.matrix.org`), sharing your IP address with us. You can also manage this in Settings.
>
>[Cancel] [Use Public Server]

**Comment by jryans** [comment]
> fwiw this could also be implemented as a dedicated SettingController to deal with the MatrixClient stuff (both reading and writing). It might not be worth the complexity though.

Ah yeah, good to know! It's probably okay for now as is I think, but if we were to sprinkle it around further, then I'd probably change to that.

**Comment by jryans** [comment]
A further revision from Matthew:

> Call failed due to misconfigured server
>
> Please ask the administrator of your homeserver (`arasphere.net`) to [configure a TURN server](https://github.com/matrix-org/synapse/blob/master/docs/turn-howto.rst) in order for calls to work reliably.
>
> Alternatively, you can try to use the public server at `turn.matrix.org`, but this will not be as reliable, and it will share your IP address with that server. You can also manage this in Settings.
>
> [OK] [Try using turn.matrix.org]

File: src/settings/Settings.js (hunk 2 above)
Comment (jryans) [reply src/settings/Settings.js]: I have updated it to be more detailed.

**Pull request merged**